src/core/expression/qgsexpression.cpp

```diff
@@ -230,7 +230,8 @@ bool QgsExpression::hasParserError() const
 
 QString QgsExpression::parserErrorString() const
 {
-  return d->mParserErrorString;
+  return d->mParserErrorString.replace( "syntax error, unexpected end of file",
+                                        tr( "Invalid expression. Unexpected end. You might not have finished the full expression" ) );
```

Comment:
```suggestion
                                        tr( "Invalid expression. Unexpected end. You might not have finished the full expression." ) );
```

Reply:
What about something like "Expression is incomplete. Are you missing a closing parentheses?". That's a little more friendly imo.

Reply:
> Are you missing a closing parenthesis?

if that's a common error, then this could be tested by just appending a parenthesis, and if re-compiling the expression and it validates, then we could be affirmative. Otherwise, if we don't known the cause of the error, the generic message is more appropriate IMHO

Reply:
I thought about the "Are you missing a closing parenthesis" message but worried it would be miss information.  The "Expression is incomplete" part sounds good though.

Reply:
Ah yeah, I guess it could be a missing ' or " too... Could we still replace "invalid expression. Unexpected end." With "incomplete expression"?

Reply:
yeah will update the wording.